[SYCL][Graph] Moving non-implemented features to extra section (#282)
reble/llvm: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc

|===

:sycl-kernel-function: https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sycl-kernel-function

Table {counter: tableNumber}. Member functions of the `command_graph` class (executable graph update).
[cols="2a,a"]
|===
|Member function|Description

|
[source, c++]
----
void

> update(const command_graph<graph_state::modifiable>& graph);

Reviewer: There's a section on `update` on line 534, do we want to move that too?

Also, line 386 in the API overview is the update declaration that can be removed.

Author: Good catch. I've (re-)moved this part on updating a graph.

=== Storage Lifetimes [[storage-lifetimes]]

The lifetime of any buffer recorded as part of a submission
to a command graph will not automatically be extended in keeping with the common reference

> semantics and buffer synchronization rules in the SYCL specification. It is a User's

Reviewer 2: ```suggestion
semantics and buffer synchronization rules in the SYCL specification. It is a user's
```

responsiblity to keep a buffer and it's underlying data alive until the buffer is no longer

> required by the graph.

Reviewer 2: Since we mention the underlying host data specifically below I don't think we need to mention it here as well.

> responsiblity to keep a buffer and it's underlying data alive until the buffer is no longer
> required by the graph.

Author: ```suggestion
responsibility to keep a buffer alive until it's no longer required by the graph.
```

Author: Section has been removed.

If a buffer created with a host data pointer is recorded as part of a submission to
a command graph, the lifetime of that host data will also not be extended.

> To illustrate this issue, consider the following example:

Reviewer 2: ```suggestion
a command graph, the lifetime of that host data will not be extended
and the user must not delete or modify that data until the associated
buffer is no longer required by the graph.
To illustrate this issue, consider the following example:
```
`also` was confusing here, and could use a bit more clarity similar to the above section.

Reviewer 2: and to be honest, I don't know if we need this section altogether. Since we no longer extend the lifetime of the buffer, if we say the user has to keep the buffer alive then they already have to keep the host data alive as well (and not modify it) as this is prohibited by the SYCL specification anyway.

Author: Okay, I don't mind removing it.

}
----

In this example, lifetime of memory might need to be extended because

> it of being used in the recorded graph. As illustrated

Reviewer 2: This example doesn't really make as much as sense now that we've removed the host copies, since it applies more to buffers in general now rather than host data explicitly.

Author: I kept it to illustrate what could go wrong, will be removed together with this section.

above, that host memory might go out of scope before the recorded graph goes out
of scope, or before the data has been copied to the device.

=== Host Tasks

:host-task: https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#subsec:interfaces.hosttasks

A {host-task}[host task] is a native C++ callable, scheduled according to SYCL

> dependency rules. It is not valid to record a host task as part of graph.

Reviewer 2: I think we should link to the planned host task information here, or somehow indicate that this is something that will change.

Author: Yes good Idea, will add a link to the specific section.

or layered as a separate extension proposal.

== Non-implemented features

> The following features are not yet implemented:

Reviewer: We don't mention subgraphs here, but we have limitations to what we currently support there, so think we should say something

From our README
> Implemented, with the limitations that a subgraph can only be added as a node to any parent graph once, and will not correctly execute by itself after being added as a sub-graph.

Author: done.

The following features are not yet implemented:

. Extending lifetime of buffers used in a graph.
. Buffer taking a copy of underlying host data when buffer is used in a graph.
. Executable graph `update()`.

> . Using `handler::host_task` in a graph node.

Reviewer: can't comment on line 1698, but backticks round `handler::memadvise`

Author: done.
